tests/tx/test_scripts.py

```diff
@@ -972,3 +982,105 @@ def test_get_sigops_count(self):
+
+        with self.assertRaises(AssertionError):
+            op_sighash_bitmask(ScriptContext(stack=[0b111, 0b101], extras=Mock(), logs=[], settings=Mock()))
+
+        stack: list[bytes | int | str] = [bytes([0b0]), bytes([0xFF, 0xFF])]
+        context = Mock(spec_set=ScriptContext)
+        extras = Mock(spec_set=ScriptExtras)
+        context.stack = stack
+        context.extras = extras
+
+        with self.assertRaises(CustomSighashModelInvalid):
+            op_sighash_bitmask(context)
+
+        context.stack = [bytes([0b111]), bytes([0b101])]
+        extras.input_index = 3
+
+        with self.assertRaises(InputNotSelectedError):
+            op_sighash_bitmask(context)
+
+        context.stack = [bytes([0b111]), bytes([0b101])]
+        extras.input_index = 2
+        op_sighash_bitmask(context)
+
+        self.assertEqual(stack, [])
+        context.set_sighash.assert_called_once_with(
+            SighashBitmask(
+                inputs=0b111,
+                outputs=0b101
+            )
+        )
+
+    def test_op_max_inputs_outputs(self) -> None:
+        context = Mock(spec_set=ScriptContext)
+        with self.assertRaises(MissingStackItems):
+            context.stack = []
+            op_max_inputs_outputs(context)
+
+        with self.assertRaises(MissingStackItems):
+            context.stack = [b'']
+            op_max_inputs_outputs(context)
+
+        with self.assertRaises(AssertionError):
+            context.stack = [1, 2]
+            op_max_inputs_outputs(context)
+
+        context.stack = [bytes([0]), bytes([0])]
+        context.extras = Mock(spec_set=ScriptExtras)
+
+        with self.assertRaises(InputsOutputsLimitModelInvalid):
+            op_max_inputs_outputs(context)
+
+        context.stack = [bytes([1]), bytes([2])]
+        context.extras.tx.inputs = ['a', 'b']
+
+        with self.assertRaises(MaxInputsExceededError):
+            op_max_inputs_outputs(context)
+
+        context.stack = [bytes([1]), bytes([2])]
+        context.extras.tx.inputs = ['a']
+        context.extras.tx.outputs = ['a', 'b', 'c']
+
+        with self.assertRaises(MaxOutputsExceededError):
+            op_max_inputs_outputs(context)
+
+        context.stack = [bytes([1]), bytes([2])]
+        context.extras.tx.inputs = ['a']
+        context.extras.tx.outputs = ['a', 'b']
+
+        op_max_inputs_outputs(context)
+
+        self.assertEqual(context.stack, [])
+
+    def test_execute_op_code(self) -> None:
+        with (
+            patch('hathor.transaction.scripts.opcode.is_opcode_valid', lambda _: False),
+            self.assertRaises(ScriptError)
+        ):
+            execute_op_code(opcode=Mock(), context=Mock())
+
+        with (
+            patch('hathor.transaction.scripts.opcode.is_opcode_valid', lambda _: True),
+            self.assertRaises(ScriptError)
+        ):
+            execute_op_code(opcode=Opcode.OP_0, context=Mock())
```

Comment:
What are we testing here?

Reply:
That when `is_opcode_valid` returns `True`, execution must fail if it's not a "function opcode". I added a comment for it in https://github.com/HathorNetwork/hathor-core/pull/912/commits/2bf75499d3e715b87646a1b5ec0f018d4655847c